src/btree/bt_sync.c

```diff
@@ -438,6 +438,17 @@ __wt_sync_file(WT_SESSION_IMPL *session, WT_CACHE_OP syncop)
                     __wt_checkpoint_progress(session, false);
             }
         }
+
+        /*
+         * During normal checkpoints, mark the tree dirty if the btree has modifications that are
+         * not visible to the checkpoint. There is a drawback in this approach as we compare the
+         * btree's maximum transaction id with the checkpoint snap_min and it is possible that this
+         * transaction may be visible to the checkpoint, but still, we mark the tree as dirty if
+         * there is a long-running transaction in the database.
+         */
+        if (!btree->modified && !F_ISSET(conn, WT_CONN_RECOVERING | WT_CONN_CLOSING_CHECKPOINT) &&
+          btree->rec_max_txn >= txn->snap_min)
```

Comment:
Checks only the max transaction id is in the checkpoint transaction snapshot or not. If we compare the max timestamp with the checkpoint timestamp, in some scenarios we are not increasing the stable timestamp and performing multiple checkpoints is leading to unnecessarily marking the tree dirty and failing a couple of our alter tests where it cannot able to close the handle due to being dirty. The transaction snapshot itself covers the scenario.

If anyone feels like considering the timestamp also, I will fix the tests by increasing the stable timestamp before performing any `alter` commands.

Reply:
I think timestamp should be included as well. I wonder if the check could be strictly greater for timestamps? Since the checkpoint timestamp is included in the checkpoint (isn't it?)

Reply:
I included the timestamp also as part of the condition to decide whether the tree can be dirty or not. The only additional change is that we do not mark the tree dirty if there is no change in the stable timestamp from the last checkpoint.

Reply:
> I wonder if the check could be strictly greater for timestamps? Since the checkpoint timestamp is included in the checkpoint (isn't it?)

Yes. The check is greater for timestamps. We include the data for the checkpoint timestamp as part of the checkpoint.